Minmin overhaul

CSharpM7 opened the pull request:
Required Assets:
[hdr-minmin-assets.zip](https://github.com/HDR-Development/HewDraw-Remix/files/11580118/hdr-minmin-assets.zip)


# Overview
This rework attempts to change Min Min by removing what players find frustrating about fighting against her at a fundamental level, while still keeping her an interesting character. Instead of a more flowchart-like playstyle, Min Min will become more of a technical character with these changes, while retaining some of her original mechanic. With the ability to throw both arms gone, she is now more in line with traditional fighters. She now sports new normals that are ideal for comboing into a punch, or at least get opponents away from her. Hopefully, this set of changes makes more players curious about trying her out. This rework contains new motion files, several status script edits (most of them being only a few lines), a new status variable for Min Min, and a single opff function

# New Normals
With the dual arms mechanic gone, this means she gains a forward tilt, neutral air, forward air, and back air. Forward Smash is temporarily the same until a new animation can be made, but expect that to be changed as well. Her grounded normals are decent at knocking opponents away from her, while her aerials are great for combing into Arms Specials.

## FTilt
![ftilt1](https://user-images.githubusercontent.com/13909643/228398417-6e5afd14-b82f-483f-ba28-6f8cf18ee68a.png)
![ftilt2](https://user-images.githubusercontent.com/13909643/228398422-8ac41d9e-cd60-42bd-9810-4cda5fdca424.png)
![ftilt3](https://user-images.githubusercontent.com/13909643/228398426-0dce6370-1e4b-4cc7-b65d-f1ffdfd7dc8f.png)
neutral angle
![ftiltHi1](https://user-images.githubusercontent.com/13909643/228398431-707899cf-cadf-488c-a935-1abcae8bc987.png)
![ftiltHi2](https://user-images.githubusercontent.com/13909643/228398434-b9b113c9-a6fc-4204-9f27-459035cdafe4.png)
![ftiltHi3](https://user-images.githubusercontent.com/13909643/228398436-99efe40a-7b05-4a1f-9d7a-83635313fa59.png)
high angle
![ftiltLw1](https://user-images.githubusercontent.com/13909643/228398438-2017ffe3-fe19-47c1-bd29-a3de87a0eb5a.png)
![ftiltLw2](https://user-images.githubusercontent.com/13909643/228398439-06da6a81-e3ad-49e4-96c4-79265a674b8b.png)
![ftiltLw3](https://user-images.githubusercontent.com/13909643/228398443-0f740f47-6baf-4f33-aea8-b6420a569f9a.png)
low angle

ZSS kick that's mainly used to get opponents away
```
- Can be angled
- Damage: 7.0/6.0 (8.0/6.5 angled)
- Angle: 361
- KBG: 80
- BKB: 50
- Active Frames: 6-8
- FAF: 29
```

## Nair
![nair1](https://user-images.githubusercontent.com/13909643/228398477-596322b5-73fb-47fa-929c-8522a3eb2281.png)
![nair2](https://user-images.githubusercontent.com/13909643/228398481-b267549e-1fa6-4f2f-b7a1-12a680fab38d.png)
![nair3](https://user-images.githubusercontent.com/13909643/228398483-348b1cc5-c6b0-4335-ad84-6fcf1a491df1.png)
![nair4](https://user-images.githubusercontent.com/13909643/228398485-70866777-7017-442c-a9af-e7b92d4801e0.png)
![nair5](https://user-images.githubusercontent.com/13909643/228398486-2438df5d-fc18-4043-8462-1c5548c283fc.png)

Standard-issue 64 kick. Great for comboing into Arms
```
- Damage: 6.0/4.0
- Angle: 361
- KBG: 67
- BKB: 37
- Active Frames: 6-8,9-25
- FAF: 45
- Landing Lag: 8
```

## Fair
![fair1](https://user-images.githubusercontent.com/13909643/228398522-7d1c148d-b400-42eb-87bc-dc59e93d4505.png)
![fair2](https://user-images.githubusercontent.com/13909643/228398527-39587790-e25d-409a-980c-084b0a1006de.png)
![fair3](https://user-images.githubusercontent.com/13909643/228398530-334f8137-a03f-4036-9424-ac93f2684a7a.png)
![fair4](https://user-images.githubusercontent.com/13909643/228398534-c93e5a27-e415-48b9-be5b-8fea3e4269f7.png)

An overhead axe kick that has a (semi)spike hitbox on her heel
```
- Damage: 12.5/10.0
- Angle: 274 (G) 34 (A) / 361
- KBG: 90/100
- BKB: 25
- Active Frames: 15-19 (16-18 Sweetspot)
- FAF: 48
- Landing Lag: 11
```

## Bair
![bair1](https://user-images.githubusercontent.com/13909643/228398550-80e6c4e7-4422-4d10-8b92-0ad7e82c5bfe.png)
![bair2](https://user-images.githubusercontent.com/13909643/228398554-306a06ec-7305-4d2f-a912-634434ffd4ad.png)

Standard issue roundhouse backair. It's not quite as potent as other back airs, but it can also combo into arms when landed
```
- Damage: 11/9
- Angle: 42
- KBG: 98
- BKB: 42
- Active Frames: 8-10
- FAF: 40
- Landing Lag: 10
```

## Standing Grab
![grab1](https://github.com/HDR-Development/HewDraw-Remix/assets/13909643/820de65a-223f-4efb-b172-0e0681dec4ff)
![grab2](https://github.com/HDR-Development/HewDraw-Remix/assets/13909643/1873cbd4-8604-4fd3-9c41-6edca7b2eb37)

Like ZSS, Minmin now has a fast standing grab
```
- Active on frames 9 and 10
- FAF: 40
```

# Arms Specials
Neutral and Side Special still retain their original properties. This also means Neutral Special in the air (Nair Special) is her original nair. The various properties of each arm have also been changed, with Ram Ram getting a large rework, Megawatt getting damage and recovery frame nerfs, and Dragon receiving significant angle nerfs, and a length nerf. Up Special (Air) has also been changed, which makes it better to finish most of her combos with while sending opponents more upward instead of at 45 degrees.

## Nair Special
```diff
+ Fixed an issue where left/right arms had different landing lag
! Ram Ram angle: 38->361
```

## Up Special (Air)
```diff
! The angle players are sent at more accurately reflects the trajectory of the arm (45->70)
+ Normal Dragon KBG Increased (78->88)
+ Power Dragon KBG Increased (85->92)
```

## Ram Ram (Arms Special)
https://user-images.githubusercontent.com/13909643/228398663-f57ed8ed-0a87-4902-bf8e-12973ce66c67.mp4

Ram Ram now rebounds opponents towards her at the apex of the arm. With a new ability to attack cancel recoil endlag, this can make Ram Ram a deadly combo starter...if you can control the hitbox properly. These buffs also come with nerfs, where Ram Ram sports a sourspot on the initial frames that sends opponents upwards. The angle control has also been heavily nerfed, which makes hitting the ledge slightly more difficult
```diff
+ Min Min can Jab,Tilt,Smash and Jump Cancel arms recoil beginning on Frame 8. Jump Cancel only works on ground
- Length Reduced (3.0/3.7->2.25/3.0)
- Time to reach apex increased by 2 frames (Smash/Charge)
- Ram Ram projectile radius decreased (12/10->10->8)
- Upwards and downwards max angle reduced
- Upwards and downwards angle acceleration reduced
! Startup hitboxes are now considered sourspots
- Sourspot Angle increased: 30->55
+ Sourspot BKB increased: 25->30
! RamRam now has a rebound hitbox (Tilt: Frame 6+ (5+ in air) Smash/Charged: Frame 4+) 
! Rebound hitboxes send grounded opponents towards MinMin, and airborne opponents upwards (to prevent ledge shenanigans)
! Rebound Angle: 50 Grounded (Backwards), 92 Airborne 
! Rebound KBG: 25/27/0 Grounded, 15 Airborne
! Rebound BKB: 60/45/90 Grounded, 80/86/86 Airborne
! Rebound Fixed Knockback on Charged version
```
## Ram Ram Projectile (Arms Special)
Ram Ram will sends aerial opponents upwards during the first few frames of the projectile to prevent ledge shenanigans. Other hitboxes will send opponents towards Min Min.
```diff
- Active Frames reduced by 3
! Has a Startup Scoop hitbox until frame 3
! Startup Angles: 50 (Backwards) Grounded, 92 Airborne
! Startup KBG: 40/0 Grounded, 90 Airborne
! Startup BKB: 60/90 Grounded, 40 Airborne
! Normal Backwards Angle: 35
! Normal KBG: 40/0
! Normal BKB: 60/90
! Fixed Knockback on charged version
```

## Megawatt (Arms Special)
Megawatt now acts more like a super smash attack. It has decent startup, and long recovery (rewind) frames, but still packs a punch. Damage was also reduced to fit a little more alongside the other arms.
```diff
- Damage reduced (Smash: 19/21->16->18; Charged: 21.85/24.15->19/21)
- (Air Only) Time to reach apex increased by 2 frames 
! BKB slightly increased to compensate damage changes (Smash,Charged)
- Active Frames reduced by 2
- Rewind speed heavily decreased
! Angle increased: 45->55
```

## Dragon (Arms Special)
Dragon is more-or-less the go-to arm now, with a decent combination of speed and knockback, without having to be as "technical" as Ram Ram can be. Dragon is also the arm that has the least controllable angles. The nerfs ensure pressing one button doesn't send an opponent towards the bottom blast zone, and also prevents mashing this arm more. The buffs help a mostly-charged smash attack link into the laser beam
```diff
- Length Reduced (3.5/4.5->3.0/3.5)
- Upwards and downwards max angle heavily reduced
- Upwards and downwards angle acceleration heavily reduced
- Angle Increased (361->45)
- Rewind speed decreased
+ New Dragon Beam Linker hitbox on Frame 6 (Charged, Grounded only)
! Linker Angle: 361
! Linker KBG: 540
! Linker Additional Hitstun: 5
! Fixed Knockback on Linker
```

CSharpM7 committed:
Init

CSharpM7 committed:
special fix

CSharpM7 committed:
Update specials.rs

CSharpM7 committed:
Update aerials.rs

CSharpM7 committed:
priority fix

CSharpM7 committed:
initial ramram translation

CSharpM7 committed:
Ramram rebound

CSharpM7 committed:
Merge branch 'HDR-Development:dev' into minmin-overhaul

CSharpM7 committed:
angle adjustments

CSharpM7 committed:
good enough

CSharpM7 committed:
Testing new rebound hitbox

CSharpM7 committed:
Ramram rebound improved, dragon angle nerf

CSharpM7 committed:
straight dragon angle

CSharpM7 committed:
ramram cancels, megawatt nerf

megawatt damage (uncharged): 19/21/19->14/16/14

CSharpM7 committed:
ramram cancel fix, megawatt slight buff

CSharpM7 committed:
Update opff.rs

CSharpM7 committed:
Several changes

UpSpecial angle changes
Jump cancel ramram
Nair angles
Megawatt frame data

CSharpM7 committed:
Speed changes

CSharpM7 committed:
Dragon AMCD and Speed

CSharpM7 committed:
Update specials.rs

CSharpM7 committed:
Code Cleanup, disabled fsmash

CSharpM7 committed:
Update aerials.rs

CSharpM7 committed:
Merge branch 'HDR-Development:dev' into minmin-overhaul

CSharpM7 committed:
Changes to the rest of the kit

## Stats
```diff
! Shorthop slightly decreased (14->12)
```

## FTilt
```diff
! BKB increased, KBG decreased
```

## Dtilt
```diff
- Sliding speed decreased by 75%
```

## UpSmash
```diff
- Reflector Hitbox (Leg) Removed
- Reflector Hitbox (Top) Active Frames Reduced to 3
```

## Nair
```diff
- FAF: 6->7
- Autocancel Start: 40->45
```

## Fair
```diff
+ Startup: 15->14
+ Sweetspot Active Frames Increased By 1
+ Sweetspot Angle: 34->1
+ Sweetspot Size Increased: 4.5->5.0
+ Damage Increased: 12.5/10->13.25->11.5
```

## Bair
```diff
+ Active Frames increased by 1
```

## Upair
```diff
- Tipman Active Frames: 11-17->12->15
- Angle Decreased: 72->70
```

## Dair
```diff
! Stall and fall removed
! Hitboxes adjusted
- Landing hitbox removed
+ Landing Lag: 20->13
- Active Frames: 14-19/19-34->14-17/17-28
+ FAF Decreased
```

# Arms Specials
## Neutral Special
```diff
- Dragon Landing Lag: 8->10
+ Dragon Damage: 8.5->9.5
+ Dragon BKB: 40->45
! Dragon Angle: 50->45
- Megawatt Landing Lag: 8->12
! Can no longer tap Special to jab
```

## Ram Ram (Arms Special)
```diff
! Sourspot sends more upwards, with more BKB, to clearly define that it is a sourspot
+ Sourspot active frames decreased to 2
+ Charged version now combos into rebound hitbox more reliably
```

## Megawatt (Arms Special)
```diff
+ Launch Speed Increased
+ Recoil Speed Increased
```

CSharpM7 commented:
![](https://user-images.githubusercontent.com/13909643/229410538-364d7172-80d0-4939-a29b-24aa04202bd4.png)

New dair hitbox for the removed stall n fall

CSharpM7 committed:
Code Cleanup

CSharpM7 committed:
Fixed damage on fair sourspots

CSharpM7 committed:
New UpB Mechanics

UpB (attack) can now be controlled up to 5 degrees back, 15 degrees forward
Tapping UpB on the ground lets you perform the Attack variant

CSharpM7 committed:
UpB Effects And Tap  Adjustments

Tap Threshold: 10 frames
Angle is now set before the arm is launched

CSharpM7 committed:
Damage Buffs and several other changes

## Jabs
```diff
- Jab 1 can no longer be held
+ Added lower jablock hitbox
! Jab 1 hitboxes adjusted
+ Jab 2 Active Frames Increased By 1 to match animation
! Hitlag normalized to 1.0
```

## Ftilt
```diff
+ Damage: 7.0/6.0->9.5/8.5 (BKB decreased slightly)
```
## Uptilt
```diff
- BKB increased: 70/65->75->70
```
## Dtilt
```diff
! BKB increased: 70->80
```

## UpSmash
KBG decreased slightly due to how stupendous her DACUS is
```diff
- KBG: 100/103->95/100
```

## Nair
```diff
+ Damage: 6.0/4.0->9.5/8.25 (BKB decreased slightly)
```

## Fair
```diff
+ Sweetspot has priority over sourspots
+ Damage: 13.25/11.5->15.0/13.75
- KBG decreased: 90->80
```

## Bair
```diff
+ Damage: 11/9.0->12.5/10.5
```
## Upair
```diff
- BKB increased: 44->49
```

## Dair
```diff
+ Late Active Frames increased by 4
+ No longer causes screenshake on landing
```

## Special Nair
```diff
+ Damage: 6.0/9.5/12.0->8.5/10.25/13.0
```

## Up Special (Attack)
Overall a very potent combo finisher
```diff
! BReverse window tightened to frame 2 when grounded, should prevent accidental b reverses
+ Damage: 9.0/10.35 -> 12.75/14.0 (KBG decreased 84->79 to slightly compensate the large damage buff)
! Hitlag increased
- Startup increased (12->14)
```

## RamRam
Angle nerfs in favor of more consistent rebounding
```diff
- Grounded Projectile Launch angle increased to prevent ledge shenanigans (30->40)
+ Aerial Projectile Launch angle decreased for stability (30->20)
- Decreased grounded downward angle range
+ Tilt Damage: 5/5->5/6
+ Rebound and Scoop boxes merged, should make rebounding more reliable while also preventing ledge shenanigans
```
## Megawatt
```diff
+ Increased angle range on ground
```
## Dragon Arm
```diff
+ Damage (Charged): 13.8/15.0/14.5 -> 13.8/16.75/16.0
- Damage (Laser): 7.0/12.0 -> 6.0/9.0
+ KBG (Laser): 70->75
! Audio Queue for when Charged Dragon is ready
! Altered Effects for Charged Dragon
```

## UpThrow (Requires romf reinstall)
```diff
+ FAF: 33->31
```

CSharpM7 commented:
![2023040413132812](https://user-images.githubusercontent.com/13909643/229909691-00f9059e-e57f-4a11-bd39-8e6901d53da0.png)
New jab1 hitbox

hdr-balance-backroom reviewed (changes requested):
- Readjust jab 1 and 2 knockback to reduce frame advantage
- Lower late dash attack angle
- Dragon arm forward smash seems to have more knockback than megawatt at later percents
- Increase up smash startup to frame 9
- Remove last two frames of activity from both up air and up tilt
- Simplify forward air's hitboxes and decrease sweetspot knockback growth by about 15-20
- Move forward air trail to align with the edge of the hitbox
- An interesting concept we thought of would be to allow arm switching during Min Min's normals (use `WorkModule::on_flag(*FIGHTER_TANTAN_INSTANCE_WORK_ID_FLAG_SPECIAL_LW_CHANGE_PUNCH_R)`)

CSharpM7 committed:
Re: Balance Backroom

Jab 1: BKB 25->20
Jab 2: BKB 30->25. Damage increased from vanilla (2.0->2.4)

Dash Attack: Late Angle: 78->72
Up Smash: Startup: 7->9
Up Air: Tipman hitbox removed. Normal hitbox active frames: 6-12->6-13
Up Tilt: Last Active Frame: 17->15. BKB increased on later hitboxes: 70->75
Fair: Foot Sourspot removed, normalized IDs, KBG on sweetspot: 85->63. Effect Trail adjusted
UpB: Attack Input cutoff: Frame 10->30
Megawatt (Hold/Charge): KBG 70->75. Damage adjustments to be more inline with rest of cast and vanilla changes (Tilt: 11->13, Hold: 16/18->17/19, Charge: 19/21->20/22)
Dragon (Charged): KBG matches Hold hitboxes
Down Special: Can now be used during normals (outside of hitstop) to switch arms

CSharpM7 committed:
code cleanup

Added 4 frames of activity to late dair

WuBoytH commented on fighters/tantan/src/opff.rs:
```diff
@@ -41,8 +41,19 @@ unsafe fn recoil_cancel(boma: &mut BattleObjectModuleAccessor,status: i32,situat
+    if [*FIGHTER_STATUS_KIND_ATTACK_S3, *FIGHTER_STATUS_KIND_ATTACK_HI3, *FIGHTER_STATUS_KIND_ATTACK_LW3, *FIGHTER_STATUS_KIND_ATTACK_AIR, *FIGHTER_STATUS_KIND_ATTACK_DASH, *FIGHTER_STATUS_KIND_ATTACK_S4, *FIGHTER_STATUS_KIND_ATTACK_HI4, *FIGHTER_STATUS_KIND_ATTACK_LW4].contains(&status_kind)
+       || ([*FIGHTER_STATUS_KIND_ATTACK].contains(&status_kind) && motion_kind == hash40("attack_13")){
+        if !boma.is_in_hitlag() {
+            if boma.is_cat_flag(Cat1::SpecialLw) {
```
Make sure to clear the command cat after setting this flag, or else it will try to keep enabling this flag and/or use the original down b swap code as well.

CSharpM7 committed:
Add Clear Commands to Arms Switch

hdr-balance-backroom reviewed (dismissed):
- Forward tilt is able to be jump canceled into an aerial at any point, unsure if this is an enlarged attack cancel window or something
- Increase down air faf from 26 to ~35-40
- Lower late dash attack angle to 40
- Decrease jab 1 and 2 frame advantage (should only be able to true combo into the next jab at relevant percents)

CSharpM7 committed:
Bug Fixes + Standing Grab

-Pressing Jump and Grab at the same time no longer results in a wavedash
-Minmin can no longer jump cancel Ftilt
-Minmin now has a standing grab akin to ZSS

CSharpM7 committed:
Bandage on JC Ftilt

SkewedAskew committed:
Merge branch 'dev' into minmin-overhaul

CSharpM7 committed:
Updated Jab 1 and 2 advantage, Normalized DA angle

- Late dash attack has an angle of 40 (same as normal hit)
- Jab 1 and Jab 2's frame advantage has been adjusted to combo at mid/mid-late percent

CSharpM7 committed:
Merge branch 'minmin-overhaul' of https://github.com/CSharpM7/HewDraw-Remix into minmin-overhaul

WuBoytH reviewed (approved):
I went to find why f tilt jump cancel exists and found the boonies so you can just have this one
